### ISSUE TITLE
Wrong formatting with bundled clang-format

### ISSUE THREAD

**Issue by FredM67** [microsoft/vscode-cpptools#11027]
### Environment

Version: 1.79.0-insider (system setup)
Commit: 667d4629dee90744edd5054f13bc2452c8c14490
Date: 2023-05-29T05:20:12.992Z
Electron: 22.5.3
Chromium: 108.0.5359.215
Node.js: 16.17.1
V8: 10.8.168.25-electron.0
OS: Windows_NT x64 10.0.22000

Published 3/30/2016, 03:27:12
Last released 5/23/2023, 02:32:56
Last updated 5/25/2023, 17:09:04
Identifier ms-vscode.cpptools

### Bug Summary and Steps to Reproduce

Bug Summary:
C++ formatting with the bundled clang-format is different from formatting with the native clang-format.
A .clang-format file is present in the current folder of the source file.

Easier to reproduce inside WSL. This way, you can easily compare the result with the native linux clang-format package.

Steps to reproduce:
1. Rightclick "Format Document With..."
2. Select "C/C++"

You get this formatting:
![image](https://github.com/microsoft/vscode-cpptools/assets/45318189/57108582-450d-4fad-b4af-11106214d996)

At step 2, Select "clangd", and you get the expected behaviour.
Expected behavior:
![image](https://github.com/microsoft/vscode-cpptools/assets/45318189/90076c5c-cb00-42bf-8690-149dbca94db3)

Formatting with the "clang-format" application under WSL (Ubuntu 22.04) or any other native Debian/Ubuntu produces the exact same result as the expected behaviour.

### Configuration and Logs

```shell
-------- Diagnostics - 5/29/2023, 9:44:20 PM
Version: 1.15.4
Current Configuration:
{
    "name": "PlatformIO",
    "includePath": [
        "/home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/include",
        "/home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp",
        "/home/metrichf/.platformio/packages/framework-arduino-avr/libraries/SoftwareSerial/src",
        "/home/metrichf/.platformio/packages/framework-arduino-avr/cores/arduino",
        "/home/metrichf/.platformio/packages/framework-arduino-avr/variants/standard",
        "/home/metrichf/.platformio/packages/framework-arduino-avr/libraries/EEPROM/src",
        "/home/metrichf/.platformio/packages/framework-arduino-avr/libraries/HID/src",
        "/home/metrichf/.platformio/packages/framework-arduino-avr/libraries/SPI/src",
        "/home/metrichf/.platformio/packages/framework-arduino-avr/libraries/Wire/src"
    ],
    "browse": {
        "limitSymbolsToIncludedHeaders": true,
        "path": [
            "/home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/include",
            "/home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp",
            "/home/metrichf/.platformio/packages/framework-arduino-avr/libraries/SoftwareSerial/src",
            "/home/metrichf/.platformio/packages/framework-arduino-avr/cores/arduino",
            "/home/metrichf/.platformio/packages/framework-arduino-avr/variants/standard",
            "/home/metrichf/.platformio/packages/framework-arduino-avr/libraries/EEPROM/src",
            "/home/metrichf/.platformio/packages/framework-arduino-avr/libraries/HID/src",
            "/home/metrichf/.platformio/packages/framework-arduino-avr/libraries/SPI/src",
            "/home/metrichf/.platformio/packages/framework-arduino-avr/libraries/Wire/src"
        ]
    },
    "defines": [
        "PROJECT_PATH=\"Mk2_3phase_RFdatalog_temp\"",
        "CURRENT_TIME=\"2023-05-29 21:25:47+02:00\"",
        "BRANCH_NAME=\"N/A\"",
        "COMMIT_HASH=\"N/A\"",
        "PLATFORMIO=60108",
        "ARDUINO_AVR_UNO",
        "F_CPU=16000000L",
        "ARDUINO_ARCH_AVR",
        "ARDUINO=10808",
        "__AVR_ATmega328P__",
        ""
    ],
    "cStandard": "c11",
    "cppStandard": "c++17",
    "compilerPath": "/home/metrichf/.platformio/packages/toolchain-atmelavr/bin/avr-gcc",
    "compilerArgs": [
        "-mmcu=atmega328p",
        ""
    ],
    "compilerPathIsExplicit": true,
    "cStandardIsExplicit": true,
    "cppStandardIsExplicit": true,
    "intelliSenseModeIsExplicit": false,
    "compilerPathInCppPropertiesJson": "/home/metrichf/.platformio/packages/toolchain-atmelavr/bin/avr-gcc",
    "mergeConfigurations": false
}
Translation Unit Mappings:
[ /home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/main.cpp ]:
    /home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h *
Translation Unit Configurations:
[ /home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/main.cpp ]:
    Process ID: 11878
    Memory Usage: 371 MB
    Compiler Path: /home/metrichf/.platformio/packages/toolchain-atmelavr/bin/avr-gcc
    Includes:
        /home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp
        /home/metrichf/.platformio/packages/framework-arduino-avr/libraries/SoftwareSerial/src
        /home/metrichf/.platformio/packages/framework-arduino-avr/cores/arduino
        /home/metrichf/.platformio/packages/framework-arduino-avr/variants/standard
        /home/metrichf/.platformio/packages/framework-arduino-avr/libraries/EEPROM/src
        /home/metrichf/.platformio/packages/framework-arduino-avr/libraries/HID/src
        /home/metrichf/.platformio/packages/framework-arduino-avr/libraries/SPI/src
        /home/metrichf/.platformio/packages/framework-arduino-avr/libraries/Wire/src
        /home/metrichf/.platformio/packages/toolchain-atmelavr/lib/gcc/avr/7.3.0/include
        /home/metrichf/.platformio/packages/toolchain-atmelavr/lib/gcc/avr/7.3.0/include-fixed
        /home/metrichf/.platformio/packages/toolchain-atmelavr/avr/include
    Defines:
        PROJECT_PATH="Mk2_3phase_RFdatalog_temp"
        CURRENT_TIME="2023-05-29 21:25:47+02:00"
        BRANCH_NAME="N/A"
        COMMIT_HASH="N/A"
        PLATFORMIO=60108
        ARDUINO_AVR_UNO
        F_CPU=16000000L
        ARDUINO_ARCH_AVR
        ARDUINO=10808
        __AVR_ATmega328P__
        
    Standard Version: c++17
    IntelliSense Mode: linux-gcc-x64
    Other Flags:
        --g++
        --gnu_version=70300
Total Memory Usage: 371 MB

------- Workspace parsing diagnostics -------
Number of files discovered (not excluded): 6123

loggingLevel: Debug
loggingLevel has changed to: Debug
LSP: cpptools/textEditorSelectionChange
LSP: cpptools/textEditorSelectionChange
LSP: cpptools/getCodeActions: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 291)
LSP: cpptools/activeDocumentChange: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h
LSP: cpptools/getInlayHints: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 292)
LSP: cpptools/getFoldingRanges: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 293)
LSP: cpptools/getCodeActions: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 294)
LSP: cpptools/resumeParsing
LSP: textDocument/hover: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 295)
LSP: cpptools/textEditorSelectionChange
LSP: cpptools/textEditorSelectionChange
LSP: cpptools/activeDocumentChange: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h
LSP: cpptools/resumeParsing
LSP: textDocument/didChange: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h
LSP: cpptools/textEditorSelectionChange
idle loop: reparsing the active document
Checking for syntax errors: /home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h
Queueing IntelliSense update for files in translation unit of: /home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/main.cpp
LSP: cpptools/getFoldingRanges: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 296)
LSP: cpptools/getCodeActions: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 297)
LSP: cpptools/getDocumentSymbols: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 298)
  tag parsing file: /home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h
LSP: cpptools/getSemanticTokens: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 299)
sending 12 changes to server
Error squiggle count: 0
Update IntelliSense time (sec): 0.304
LSP: cpptools/getFoldingRanges: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 300)
LSP: textDocument/willSaveWaitUntil: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 301)
willSaveWaitUntil: 0ms
LSP: textDocument/didSave: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h
LSP: cpptools/getCodeActions: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 302)
  tag parsing file: /home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h
LSP: cpptools/fileChanged: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h
LSP: cpptools/getInlayHints: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 303)
LSP: textDocument/hover: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 304)
Checking for syntax errors: /home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h
Queueing IntelliSense update for files in translation unit of: /home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/main.cpp
idle loop: reparsing the active document
Checking for syntax errors: /home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h
Queueing IntelliSense update for files in translation unit of: /home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/main.cpp
LSP: textDocument/hover: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 305)
Error squiggle count: 0
Update IntelliSense time (sec): 0.508
LSP: cpptools/getSemanticTokens: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 306)
LSP: cpptools/getFoldingRanges: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 307)
LSP: cpptools/textEditorSelectionChange
LSP: cpptools/getCodeActions: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 308)
LSP: cpptools/formatDocument: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 309)
Formatting document: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h
Formatting Engine: clangFormat
LSP: textDocument/didChange: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h
LSP: cpptools/textEditorSelectionChange
LSP: cpptools/getFoldingRanges: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 310)
LSP: cpptools/getCodeActions: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 311)
LSP: cpptools/getDocumentSymbols: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 312)
  tag parsing file: /home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h
LSP: cpptools/getSemanticTokens: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 313)
LSP: textDocument/willSaveWaitUntil: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 314)
willSaveWaitUntil: 0ms
LSP: textDocument/didSave: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h
LSP: cpptools/getCodeActions: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 315)
  tag parsing file: /home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h
LSP: cpptools/fileChanged: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h
LSP: textDocument/hover: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 316)
sending 20 changes to server
LSP: cpptools/getInlayHints: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 317)
idle loop: reparsing the active document
Checking for syntax errors: /home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h
Queueing IntelliSense update for files in translation unit of: /home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/main.cpp
LSP: textDocument/hover: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 318)
Checking for syntax errors: /home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h
Queueing IntelliSense update for files in translation unit of: /home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/main.cpp
LSP: textDocument/hover: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 319)
Error squiggle count: 0
Update IntelliSense time (sec): 0.509
LSP: cpptools/getFoldingRanges: file:///home/metrichf/git/PVRouter-3-phase/Mk2_3phase_RFdatalog_temp/config.h (id: 320)
LSP: cpptools/resumeParsing
```

### Other Extensions

Issue is identical whenever clangd extension is enable or not.

**Comment by Colengms** [comment]
Hi @FredM67 .  The C/C++ Extension isn't doing much other than invoke `clang-format` under the hood.  Is it possible that the difference you are seeing is between versions of `clang-format`?  The `clang-format` binary installed by the extension would be in your user directory, under i.e. `.vscode/extensions/ms-vscode.cpptools-1.15.4/LLVM/bin`.  You can execute it with `--version` to see the version number.

Could you try setting `C_Cpp.clang_format_path` to the path of one of your 'known good' versions of `clang-format`, and see if the result is what you expect?  

If the result of the format operation is not what you expect, even using your preferred version of `clang-format`, it's possible the issue has something to do with other formatting related settings.  Could you try setting the C/C++ extension's logging level to (the undocumented level) 7, like so? :

```
"C_Cpp.loggingLevel": "7"
```

And looking for the specific command line used to invoke `clang-format` in the log output?  That might provide the necessary clues as to what is different.

**Comment by pgerell** [comment]
This seems to be a regression in clang-format. https://github.com/llvm/llvm-project/issues/62161

**Comment by Colengms** [comment]
Hi @pgerell .  Thanks for tracking that down.  We'll close this issue as external.

**Comment by sean-mcmanus** [comment]
@pgerell It looks like this bug was fixed with 16.0.3 (we ship 16.0.0) and 16.0.5 is currently available, so we could potentially update to that. We weren't planning to update clang-format until 17.0.0, but I don't know when that is planned, so it might be many months.

**Comment by Fady2007** [comment]
compiler shows errors but visual studio code editor does not show any basic errors in syntax

For Example:

#include <iostream>
 using namespace std;
 int main() 
{ 
cout << "Line One\n";
 cout << "Line Two\n"
  cout << "Line Three\n";
 return 0;
 this is a comment with out "//" 
}

NO Formatting or shows syntax error by editor

How can i solve this problem ?

**Comment by FredM67** [comment]
How is the state ?
Will this issue be fixed "now" or stays until clang-format 17 is out?

**Comment by sean-mcmanus** [comment]
@FredM67 Our current plan is to wait for clang-format 17 -- you can workaround the issue via installing clang-format 16.0.6 and updating your C_Cpp.clang_format_path to reference it.

**Comment by sean-mcmanus** [comment]
@FredM67 This fixed now in https://github.com/microsoft/vscode-cpptools/releases/tag/v1.17.1 (although it's not mentioned in the changelog).

**Comment by FredM67** [comment]
> @FredM67 This fixed now in 1.17.1 (pre-release) https://github.com/microsoft/vscode-cpptools/releases/tag/v1.17.1 (although it's not mentioned in the changelog). It should be in the final 1.17.2 release on Monday or so.

Fine... I'll try tomorrow.. Thx

**Comment by michelleangela** [comment]
Fixed in release version https://github.com/microsoft/vscode-cpptools/releases/tag/v1.17.2.

Note: this did not get included in changelog.
